pkg/spdk/engine.go

```diff
@@ -498,6 +528,223 @@ func (e *Engine) validateAndUpdateReplicaMode(replicaName string, bdev *spdktype
+	e.ReplicaAddressMap[replicaName] = replicaAddress
+	e.ReplicaBdevNameMap[replicaName] = ""
+	e.ReplicaModeMap[replicaName] = types.ModeWO
+
+	return nil
+}
+
+func (e *Engine) ReplicaAddFinish(spdkClient *spdkclient.Client, replicaName, replicaAddress string, localReplicaLvsNameMap map[string]string) (err error) {
+	updateRequired := false
+
+	e.Lock()
+	defer func() {
+		e.Unlock()
+
+		if updateRequired {
+			e.UpdateCh <- nil
+		}
+	}()
+
+	// Syncing with the SPDK TGT server only when the engine is running.
+	if e.State != types.InstanceStateRunning {
+		return fmt.Errorf("invalid state %v for engine %s replica %s add finish", e.State, e.Name, replicaName)
+	}
+
+	if e.Frontend != types.FrontendSPDKVoid {
+		return fmt.Errorf("invalid frontend %v for engine %s replica %s add finish", e.Frontend, e.Name, replicaName)
+	}
+
+	if _, exists := e.ReplicaAddressMap[replicaName]; !exists {
+		return fmt.Errorf("replica %s does not exist in engine %s", replicaName, e.Name)
+	}
+
+	if e.ReplicaModeMap[replicaName] != types.ModeWO {
+		return fmt.Errorf("invalid mode %s for engine %s replica %s add finish", e.ReplicaModeMap[replicaName], e.Name, replicaName)
+	}
+
+	replicaBdevList := []string{}
+	for rName, rMode := range e.ReplicaModeMap {
+		if rMode == types.ModeRW {
+			replicaBdevList = append(replicaBdevList, e.ReplicaBdevNameMap[rName])
+		}
+		if rName == replicaName {
+			bdevName, err := getBdevNameForReplica(spdkClient, localReplicaLvsNameMap, replicaName, replicaAddress, e.IP)
+			if err != nil {
+				e.log.WithError(err).Errorf("Failed to get bdev from rebuilding replica %s with address %s, will mark it as ERR and error out", replicaName, replicaAddress)
+				e.ReplicaModeMap[replicaName] = types.ModeERR
+				e.ReplicaBdevNameMap[replicaName] = ""
+				updateRequired = true
+				return err
+			}
+			e.ReplicaModeMap[replicaName] = types.ModeRW
+			e.ReplicaBdevNameMap[replicaName] = bdevName
+			replicaBdevList = append(replicaBdevList, bdevName)
+			updateRequired = true
+		}
+	}
+
+	defer func() {
+		if err != nil && e.State != types.InstanceStateError {
+			e.State = types.InstanceStateError
+			updateRequired = true
+		}
+	}()
+
+	if _, err := spdkClient.BdevRaidDelete(e.Name); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+		return err
+	}
+	if _, err := spdkClient.BdevRaidCreate(e.Name, spdktypes.BdevRaidLevel1, 0, replicaBdevList); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *Engine) ReplicaShallowCopy(destReplicaName, destReplicaAddress string) (err error) {
+	e.Lock()
+	defer e.Unlock()
+
+	// Syncing with the SPDK TGT server only when the engine is running.
+	if e.State != types.InstanceStateRunning {
+		return fmt.Errorf("invalid state %v for engine %s replica %s shallow copy", e.State, e.Name, destReplicaName)
+	}
+
+	if e.Frontend != types.FrontendSPDKVoid {
+		return fmt.Errorf("invalid frontend %v for engine %s replica %s shallow copy", e.Frontend, e.Name, destReplicaName)
```

Comment:
Should it be `FrontendEmpty`?

Reply:
The PR for `types. FrontendSPDKVoid` is already merged hence I directly use this const. Do we need to rename it to `FrontendEmpty`

Reply:
I'm fine with `FrontendEmpty`. We can rename it.